addons/stock/models/stock_location.py

```diff
@@ -221,7 +221,8 @@ def _get_putaway_strategy(self, product, quantity=0, package=None, packaging=Non
 
         # get current product qty (qty in current quants and future qty on assigned ml) of all child locations
         qty_by_location = defaultdict(lambda: 0)
-        locations = self.env['stock.location'].search([('id', 'child_of', self.id), ('usage', '=', 'internal')])
+        if locations is None:
+            locations = self.env['stock.location'].search([('id', 'child_of', self.id), ('usage', '=', 'internal')])
```

Comment:
Maybe it is cleaner (and faster) to make a one2many compute on `stock.location` with the result of this search (`child_internal_location_ids` by example). It will allow to cache it (more efficiently because it can be share between different call of `_action_assign` and `_set_quantity_done_prepare_vals`) and avoid to complicate the code with optional parameters.
(Also this new field can be used also in `_get_putaway_location`)

What do you think ?

Reply:
It's interesting idea. I made a separate PR: https://github.com/odoo/odoo/pull/74016
The only problem is invalidating cache, because we can't specify descendants of  descendants  in `api.depends` for such field

Reply:
> The only problem is invalidating cache, because we can't specify descendants of descendants in api.depends for such field

Are you sure? with a recursive=True in the fields declaration and with a recursive depends ?

Reply:
Apparently, I didn't know about recursive dependencies. I'll do as you described there https://github.com/odoo/odoo/pull/74016/files#r680699773